### PR TITLE
Lucas/state reader

### PR DESCRIPTION
**Pull request by LucasLvy**
# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: 

# What is the new behavior?

- We don't copy the whole substrate state in a blockifier type anymore

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Other information

### CODE REVIEW

File: crates/pallets/starknet/src/tests.rs [hunk 1]
```diff
@@ -494,86 +491,117 @@ fn given_erc20_transfer_when_invoke_then_it_works() {
...
         run_to_block(1);
         let origin = RuntimeOrigin::none();
-        let (sender_account, _, _) = account_helper(TEST_ACCOUNT_SALT);
-        // Declare ERC20 contract
-        declare_erc20(origin.clone(), sender_account);
+        let sender_account = <[u8; 32]>::from_hex("000000000000000000000000000000000000000000000000000000000000000F").unwrap();
+        // ERC20 is already declared for the fees.
         // Deploy ERC20 contract
-        deploy_erc20(origin.clone(), sender_account);
+        let deploy_transaction = Transaction {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/tests.rs:497]: why not use `transaction_from_json` ?

File: crates/pallets/starknet/src/lib.rs [hunk 2]
```diff
@@ -1293,4 +1217,68 @@ pub mod pallet {
...
+    impl<T: Config> Default for Substate<T> {
+        fn default() -> Self {
+            Self(PhantomData)
+        }
+    }
+
+    impl<T: Config> StateReader for Substate<T> {
+        fn get_storage_at(&mut self, contract_address: ContractAddress, key: StorageKey) -> StateResult<StarkFelt> {
+            Ok(StarkFelt::new(Pallet::<T>::storage((contract_address.0.0.0, H256::from_slice(&key.0.0.0))).into())
+                .unwrap())
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1230]: remove `unwrap`

File: crates/pallets/starknet/src/lib.rs [hunk 3]
```diff
@@ -1293,4 +1217,68 @@ pub mod pallet {
             Pallet::<T>::transfer_fees(payer, to, (actual_fee + tip).as_u128())
         }
     }
+    pub struct Substate<T: Config>(PhantomData<T>);
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1220]: some docs and rationale about `Substate`

File: crates/pallets/starknet/src/lib.rs [hunk 4]
```diff
@@ -1293,4 +1217,68 @@ pub mod pallet {
...
+
+    impl<T: Config> StateReader for Substate<T> {
+        fn get_storage_at(&mut self, contract_address: ContractAddress, key: StorageKey) -> StateResult<StarkFelt> {
+            Ok(StarkFelt::new(Pallet::<T>::storage((contract_address.0.0.0, H256::from_slice(&key.0.0.0))).into())
+                .unwrap())
+        }
+        /// Returns the nonce of the given contract instance.
+        /// Default: 0 for an uninitialized contract address.
+        fn get_nonce_at(&mut self, contract_address: ContractAddress) -> StateResult<Nonce> {
+            Ok(Nonce(StarkFelt::new(Pallet::<T>::nonce(contract_address.0.0.0).into()).unwrap()))
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1235]: `unwrap`

File: crates/pallets/starknet/src/lib.rs [hunk 5]
```diff
@@ -1293,4 +1217,68 @@ pub mod pallet {
...
+        }
+
+        /// Returns the class hash of the contract class at the given contract instance.
+        /// Default: 0 (uninitialized class hash) for an uninitialized contract address.
+        fn get_class_hash_at(&mut self, contract_address: ContractAddress) -> StateResult<ClassHash> {
+            Ok(ClassHash(StarkFelt::new(Pallet::<T>::contract_class_hash_by_address(contract_address.0.0.0)).unwrap()))
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1241]: same

File: crates/pallets/starknet/src/lib.rs [hunk 6]
```diff
@@ -1293,4 +1217,68 @@ pub mod pallet {
...
+            Ok(())
+        }
+
+        /// Sets the given contract class under the given class hash.
+        fn set_contract_class(&mut self, class_hash: &ClassHash, contract_class: ContractClass) -> StateResult<()> {
+            self::ContractClasses::<T>::insert(class_hash.0.0, ContractClassWrapper::from(contract_class));
+            Ok(())
+        }
+
+        fn to_state_diff(&self) -> StateDiff {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1280]: not implemented ?

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left few comments

File: crates/pallets/starknet/src/tests.rs (hunk 1 above)
Comment (LucasLvy) [reply crates/pallets/starknet/src/tests.rs:497]: Easier to assert the event and easier to modify. Imo tx from json should be used for real txs that we take from mainnet/testnet

File: crates/pallets/starknet/src/lib.rs (hunk 6 above)
Comment (LucasLvy) [reply crates/pallets/starknet/src/lib.rs:1280]: What’s the point ?

File: crates/pallets/starknet/src/tests.rs (hunk 1 above)
Comment (EvolveArt) [reply crates/pallets/starknet/src/tests.rs:497]: Makes sense

File: crates/pallets/starknet/src/lib.rs (hunk 6 above)
Comment (EvolveArt) [reply crates/pallets/starknet/src/lib.rs:1280]: `apply_state_diffs` ?

File: crates/pallets/starknet/src/lib.rs [hunk 7]
```diff
@@ -515,12 +520,13 @@ pub mod pallet {
...
             // Execute transaction
-            match transaction.execute(state, block, TxType::DeclareTx, Some(contract_class.clone()), fee_token_address)
-            {
+            match transaction.execute(state, block, TxType::DeclareTx, Some(contract_class), fee_token_address) {
                 Ok(_) => {
                     log!(debug, "Declare Transaction executed successfully.");
                 }
                 Err(e) => {
+                    #[cfg(feature = "std")]
+                    std::println!("here {:?}", e);
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:529]: remove print

File: crates/pallets/starknet/src/lib.rs [hunk 8]
```diff
@@ -1216,10 +1140,10 @@ pub mod pallet {
                     return Err(TransactionValidityError::Invalid(Payment));
                 }
             }
-            Pallet::<T>::apply_state_diffs(state).map_err(|_| {
-                log!(error, "Couldn't apply the state diffs");
-                TransactionValidityError::Unknown(Custom(3_u8))
-            })?;
+            // Pallet::<T>::apply_state_diffs(state).map_err(|_| {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/lib.rs:1143]: remove not just comment

File: crates/pallets/starknet/src/lib.rs (hunk 6 above)
Comment (EvolveArt) [reply crates/pallets/starknet/src/lib.rs:1280]: Ok well as the state diffs are applied iteratively on the state directly we don't need it anymore, gotcha.

**Review by EvolveArt — CHANGES_REQUESTED** [review]

**Review by EvolveArt — APPROVED** [review]
lgtm

**Pull request merged**